src/librustc_resolve/lib.rs

```diff
@@ -2077,40 +2086,42 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
             this.visit_generics(generics);
```

Comment:
(alternatively, maybe we should first visit the `generics.ty_params` outside of the `with_current_self_type`, and then just move the visit of the `generics.where_clause.predicates` into the inside of `with_current_self_type`.)